models/connection/session_manager.go

```diff
@@ -0,0 +1,217 @@
+package connection
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+
+	cerr "github.com/saeidalz13/battleship-backend/internal/error"
+)
+
+type SessionManager interface {
+	GenerateNewSession(conn *websocket.Conn) *Session
+	CleanupPeriodically()
+	FindSession(sessionId string) (*Session, error)
+	TerminateSession(session *Session)
+	ReconnectSession(session *Session, conn *websocket.Conn)
+	Communicate(senderSessionId, receiverSessionId string, msg interface{}, msgType uint8) error
+	HandleAbnormalClosureSession(session *Session, otherSessionId string) error
+	GetSessionId(session *Session) string
+	WriteToSessionConn(session *Session, msg interface{}, msgType uint8, otherSessonId string) error
+	FetchCodeFromMsg(session *Session, payload []byte) (uint8, error)
+	ReadFromSessionConn(session *Session, otherSessionId string) (int, []byte, error)
+}
+
+type BattleshipSessionManager struct {
+	cleanupInterval time.Duration
+	sessions        map[string]*Session
+	mu              sync.RWMutex
+}
+
+func NewBattleshipSessionManager() *BattleshipSessionManager {
+	initMapSize := 10
```

Comment:
what is the point of this?

Reply:
When you initialize a size map, it avoids copying the underlying slice for the map.
This is irrelevant for now but in a large scale work, we can set this to something like 50! so that it would avoid copying.